src/server/auth/server/api_server.go

```diff
@@ -1175,6 +1195,252 @@ func (a *apiServer) RevokeAuthToken(ctx context.Context, req *authclient.RevokeA
 	return &authclient.RevokeAuthTokenResponse{}, nil
 }
 
+func (a *apiServer) SetGroupsForUser(ctx context.Context, req *authclient.SetGroupsForUserRequest) (resp *authclient.SetGroupsForUserResponse, retErr error) {
+	a.LogReq(req)
+	defer func(start time.Time) { a.LogResp(req, resp, retErr, time.Since(start)) }(time.Now())
+	if !a.isActivated() {
+		return nil, authclient.ErrNotActivated
+	}
+
+	callerInfo, err := a.getAuthenticatedUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if !a.isAdmin(callerInfo.Subject) {
+		return nil, &authclient.ErrNotAuthorized{
+			Subject: callerInfo.Subject,
+			AdminOp: "SetGroupsForUser",
+		}
+	}
+
+	username, err := lenientCanonicalizeSubject(ctx, req.Username)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err := col.NewSTM(ctx, a.etcdClient, func(stm col.STM) error {
+		members := a.members.ReadWrite(stm)
+
+		// Get groups to remove/add user from/to
+		var removeGroups authclient.Groups
+		addGroups := addToSet(nil, req.Groups...)
+		if err := members.Get(username, &removeGroups); err == nil {
+			for _, group := range req.Groups {
+				if _, ok := removeGroups.Groups[group]; ok {
+					removeGroups.Groups = removeFromSet(removeGroups.Groups, group)
+					addGroups = removeFromSet(addGroups, group)
+				}
+			}
+		}
+
+		// Set groups for user
+		if err := members.Put(username, &authclient.Groups{
+			Groups: addToSet(nil, req.Groups...),
+		}); err != nil {
+			return err
+		}
+
+		// Remove user from previous groups
+		groups := a.groups.ReadWrite(stm)
+		var membersProto authclient.Users
+		for group := range removeGroups.Groups {
+			if err := groups.Upsert(group, &membersProto, func() error {
+				membersProto.Usernames = removeFromSet(membersProto.Usernames, username)
+				return nil
+			}); err != nil {
+				return err
+			}
+		}
+
+		// Add user to new groups
+		for group := range addGroups {
+			if err := groups.Upsert(group, &membersProto, func() error {
+				membersProto.Usernames = addToSet(membersProto.Usernames, username)
+				return nil
+			}); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return &authclient.SetGroupsForUserResponse{}, nil
+}
+
+func (a *apiServer) ModifyMembers(ctx context.Context, req *authclient.ModifyMembersRequest) (resp *authclient.ModifyMembersResponse, retErr error) {
+	a.LogReq(req)
+	defer func(start time.Time) { a.LogResp(req, resp, retErr, time.Since(start)) }(time.Now())
+	if !a.isActivated() {
+		return nil, authclient.ErrNotActivated
+	}
+
+	callerInfo, err := a.getAuthenticatedUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if !a.isAdmin(callerInfo.Subject) {
+		return nil, &authclient.ErrNotAuthorized{
+			Subject: callerInfo.Subject,
+			AdminOp: "ModifyMembers",
+		}
+	}
+
+	add, err := lenientCanonicalizeSubjects(ctx, req.Add)
+	if err != nil {
+		return nil, err
+	}
+	remove, err := lenientCanonicalizeSubjects(ctx, req.Remove)
```

Comment:
Could we add a todo to only canonicalize subjects that we can't find (i.e. if the user gives us a username to remove and it's right there in the group, skip canonicalization and just remove it. Otherwise, canonicalize the username and then see if we still can't find it)

Reply:
Sure.